src/core/server/status/plugins_status.ts

```diff
@@ -0,0 +1,104 @@
+import { map, distinctUntilChanged, switchMap, debounce } from 'rxjs/operators';
+import { isDeepStrictEqual } from 'util';
+
+import { PluginName } from '../plugins';
+import { ServiceStatus, CoreStatus } from './types';
+import { getSummaryStatus } from './get_summary_status';
+
+interface Deps {
+  core$: Observable<CoreStatus>;
+  pluginDependencies: ReadonlyMap<PluginName, PluginName[]>;
+}
+
+const waitForMicrotasks = (numberOfTicks: number = 1): Promise<unknown> => {
+  if (numberOfTicks === 1) {
+    return Promise.resolve();
+  } else {
+    return Promise.resolve().then(() => waitForMicrotasks(numberOfTicks - 1));
+  }
+};
+
+export class PluginsStatusService {
+  private readonly pluginStatuses = new Map<PluginName, Observable<ServiceStatus>>();
+  private readonly update$ = new BehaviorSubject(true);
+  constructor(private readonly deps: Deps) {}
+
+  public set(plugin: PluginName, status$: Observable<ServiceStatus>) {
+    this.pluginStatuses.set(plugin, status$);
+    this.update$.next(true); // trigger all existing Observables to update from the new source Observable
+  }
+
+  public getAll$(): Observable<Record<PluginName, ServiceStatus>> {
+    return this.getPluginStatuses$([...this.deps.pluginDependencies.keys()]);
+  }
+
+  public getPlugins$(plugin: PluginName): Observable<Record<PluginName, ServiceStatus>> {
+    const dependencies = this.deps.pluginDependencies.get(plugin);
+    if (!dependencies) {
+      throw new Error(`Unknown plugin: ${plugin}`);
+    }
+
+    return this.getPluginStatuses$(dependencies);
+  }
+
+  public getDerivedStatus$(plugin: PluginName): Observable<ServiceStatus> {
+    return combineLatest(this.deps.core$, this.getPlugins$(plugin)).pipe(
+      map(([coreStatus, pluginStatuses]) => {
+        return getSummaryStatus(
+          [...Object.entries(coreStatus), ...Object.entries(pluginStatuses)],
+          {
+            allAvailableSummary: `All dependencies are available`,
+          }
+        );
+      })
+    );
+  }
+
+  private getPluginStatuses$(plugins: PluginName[]): Observable<Record<PluginName, ServiceStatus>> {
+    if (plugins.length === 0) {
+      return of({});
+    }
+
+    return this.update$.pipe(
+      switchMap(() => {
+        const pluginStatuses = plugins
+          .map(
+            (depName) =>
+              [depName, this.pluginStatuses.get(depName) ?? this.getDerivedStatus$(depName)] as [
+                PluginName,
+                Observable<ServiceStatus>
+              ]
+          )
+          .map(([pName, status$]) => status$.pipe(map((status) => ({ [pName]: status }))));
+
+        return combineLatest(...pluginStatuses).pipe(
+          // We schedule the number of microtasks expected to resolve all of the dependencies of this update.
+          // Waiting for this ensures that we do not emit 'partial updates' to reduce noise.
+          debounce((statuses) => waitForMicrotasks(Object.keys(statuses).length)),
```

Comment:
so it will work as expected with an assumption that plugins don't use `async` functions to update their statuses?
```js
core.status.set(
  combineLatest(
    core.status.inherited$,
    myExternalDependency$
  ).pipe(
    switchMap(async (...) => {
      await ....
      return status;
    });
  )
);
```

Reply:
Yes, that is correct. My goal was to make it a good default that works for most cases (the default derived status + any custom statuses that internal to Kibana). Do you think we should just use a time-based debounce instead?

Reply:
Going to use time-based debounce instead: https://github.com/elastic/kibana/pull/75819#discussion_r477619443